### PR TITLE
feat(core): move all types from "assets" to "core"

### PR DESCRIPTION
**Pull request by eladb**
### Commit Message
feat(core): move all types from "assets" to "core" (#7708)

Fold the "assets" module, which includes the `Staging` construct that takes care of staging asset files into the cloud assembly during synthesis into "core". This is in order to allow implementing custom resources that leverage assets throughout the framework.

NOTE: since interfaces in @aws-cdk/assets are used as based types for stable APIs, we are unable to fully deprecate this module within the codebase, so these types are left in tact and a compatibility layer was added.

A subsequent commit will add a mini-framework for custom resources that leverages this capability.
### End Commit Message

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by jogold** [comment]
> A subsequent commit will add a mini-framework for custom resources that leverages this capability.

Curious about this mini-framework...

File: packages/@aws-cdk/core/lib/fs/index.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
+import { copyDirectory } from './copy';
+import { fingerprint } from './fingerprint';
+import { CopyOptions, FingerprintOptions } from './options';
+
+export * from './options';
+
+/**
+ * File system utilities.
+ */
+export class FileSystem {
```
Comment (nija-at) [inline packages/@aws-cdk/core/lib/fs/index.ts:10]: Thanks for throwing this into a class.

File: packages/@aws-cdk/assets/lib/fs/follow-mode.ts [hunk 2]
```diff
@@ -1,3 +1,7 @@
+/**
+ * Symlink follow mode.
+ * @deprecated see `core.SymlinkFollowMode`
+ */
 export enum FollowMode {
```
Comment (nija-at) [inline packages/@aws-cdk/assets/lib/fs/follow-mode.ts:5]: Why not simply type alias or extend `FollowMode`, `CopyOptions` and `FingerprintOptions` with new ones added in core?

File: packages/@aws-cdk/core/test/test.staging.ts [hunk 3]
```diff
@@ -0,0 +1,77 @@
+import * as cxapi from '@aws-cdk/cx-api';
```
Comment (nija-at) [inline packages/@aws-cdk/core/test/test.staging.ts:1]: Why is this also not a move?

File: packages/@aws-cdk/assets/lib/fs/follow-mode.ts (hunk 2 above)
Comment (eladb) [reply packages/@aws-cdk/assets/lib/fs/follow-mode.ts:5]: Aliases are not respected by jsii

File: packages/@aws-cdk/core/test/test.staging.ts (hunk 3 above)
Comment (eladb) [reply packages/@aws-cdk/core/test/test.staging.ts:1]: Intentionally left the tests in the original module to ensure we maintain backwards compatibility.

**Pull request merged**